lib/cretonne/src/loop_analysis.rs

```diff
@@ -129,77 +169,162 @@ impl LoopAnalysis {
+                                    // entirely in the parent loop, and the one we created by
+                                    // splitting is part of only one loop, the parent loop.
+                                    if func.layout
+                                           .last_inst(ebb)
+                                           .map_or(false, |ebb_last_inst| {
+                                        ebb_last_inst != loop_edge_inst
+                                    }) {
+                                        // This handles the second case
+                                        self.split_ebb_containing_two_loops(ebb,
+                                                                            loop_edge_inst,
+                                                                            lp,
+                                                                            func,
+                                                                            domtree,
+                                                                            cfg);
+                                    }
                                 }
                             }
                         }
                     }
                 }
-                // Now we have handled the popped node and need to continue the DFS by adding the
-                // predecessors of that node
+                // Now we have handled the popped Ebb and need to continue the DFS by adding the
+                // predecessors of that Ebb
                 if let Some(continue_dfs) = continue_dfs {
-                    for &(pred, _) in cfg.get_predecessors(continue_dfs) {
-                        stack.push(pred)
+                    for &(pre, pre_inst) in cfg.get_predecessors(continue_dfs) {
+                        stack.push((pre, pre_inst))
                     }
                 }
             }
-
         }
     }
+
+    // We are in the case where `ebb` belongs partially to two different loops, the child and
+    // the parent. `lp` here is the parent loop and we create a new Ebb so that `ebb` belongs
+    // in its entirety to the parent loop.
+    // We also update the cfd and domtree to reflect that.
+    fn split_ebb_containing_two_loops(&mut self,
+                                      ebb: Ebb,
+                                      split_inst: Inst,
+                                      lp: Loop,
+                                      func: &mut Function,
+                                      domtree: &mut DominatorTree,
+                                      cfg: &mut ControlFlowGraph) {
+        let new_ebb = func.dfg.make_ebb();
+        let real_split_inst = {
+            let mut cur = Cursor::new(&mut func.layout);
+            cur.goto_inst(split_inst);
+            cur.next_inst()
+                .expect("you cannot split at the last instruction")
+        };
```

Comment:
There already is a `Layout::next_ebb()` method. We should have a `next_inst` too.